Aws fargate pricing

Tomerfi1210 opened the pull request:
What does this PR change?
Downloads and parses AmazonECS pricing data to correctly populate Amazon Fargate nodes pricing on EKS.
Does this PR relate to any other PRs?
No as far as I'm aware of
How will this PR impact users?
No action is required on user-side. Fargate nodes will now get correct pricing.
Does this PR address any GitHub or Zendesk issues?
Closes https://github.com/opencost/opencost/issues/1622
How was this PR tested?
I've published snapshot images at [ghcr.io/motoki317/opencost](https://github.com/motoki317/opencost/pkgs/container/opencost) and tested it on a real environment, and it is working fine at least for my cluster.
patch-eks-fargate includes this PR, and patch-eks includes both this PR and https://github.com/opencost/opencost/pull/3348.
Prometheus metrics node_total_hourly_cost, node_cpu_hourly_cost, and node_ram_hourly_cost match the actual fargate node cost.
For example, in ap-northeast-1, node_cpu_hourly_cost is 0.05056 and node_ram_hourly_cost is 0.00553.
If we have a Fargate node with ProvisionedCapacity of 0.25vCPU 1GB, we have node_total_hourly_cost equal to 0.05056 * 0.25 + 0.00553 * 1 = 0.01817.
https://aws.amazon.com/fargate/pricing/


NOTE: this is a forked with missing tests by @motoki317 of this PR: https://github.com/opencost/opencost/pull/3350

motoki317 committed:
feat(aws): Use AmazonECS price list to calculate Fargate nodes price

Signed-off-by: motoki317 <motoki317@gmail.com>
Signed-off-by: Tomer Fisher <tomer.fisher@zesty.co>

motoki317 committed:
fix: Stop overwriting RAMBytes if provider has filled in field

Signed-off-by: motoki317 <motoki317@gmail.com>
Signed-off-by: Tomer Fisher <tomer.fisher@zesty.co>

Tomerfi1210 committed:
feat(aws): Add tests for Fargate pricing initialization and node identification

Signed-off-by: Tomer Fisher <tomer.fisher@zesty.co>

ameijer commented:
@Tomerfi1210 looks like build failures? see the build logs above

ameijer commented on pkg/cloud/aws/provider.go:
```diff
@@ -1413,6 +1453,72 @@ func (aws *AWS) createNode(terms *AWSProductTerms, usageType string, k models.Ke
+var fargatePodCapacityRegex = regexp.MustCompile("^([0-9.]+)vCPU ([0-9.]+)GB$")
+
+func (aws *AWS) createFargateNode(awsKey *awsKey, usageType string) (*models.Node, models.PricingMetadata, error) {
+	pod, ok := aws.getFargatePod(awsKey)
```
```suggestion
    if aws.FargatePricing == nil {
      return nil, models.PricingMetadata{}, fmt.Errorf("fargate pricing not initialized")
    }
	pod, ok := aws.getFargatePod(awsKey)
```

ameijer replied:
to protect you if fargate pricing is nil

Tomerfi1210 replied:
Done.

ameijer commented on pkg/cloud/aws/fargate.go:
```diff
@@ -0,0 +1,162 @@
+	url := f.getPricingURL(nodeList)
+
+	log.Infof("Downloading Fargate pricing data from %s", url)
+	resp, err := http.Get(url)
```
please close body

Tomerfi1210 replied:
done.

ameijer commented on pkg/cloud/aws/fargate.go:
```diff
@@ -0,0 +1,162 @@
+
+	log.Infof("Downloading Fargate pricing data from %s", url)
+	resp, err := http.Get(url)
+	if err != nil {
```
check for 200 status code

Tomerfi1210 replied:
done.

ameijer commented on pkg/env/costmodel.go:
```diff
@@ -189,6 +190,11 @@ func GetAWSPricingURL() string {
 
+// GetAWSECSPricingURL returns an optional alternative URL to fetch AmazonECS pricing data from; for use in airgapped environments
+func GetAWSECSPricingURL() string {
+	return env.Get(AWSECSPricingURL, "")
```
add _OVERRIDE to this to make it clear that this is an override of the default?

Tomerfi1210 replied:
done.

ameijer commented on pkg/cloud/aws/provider.go:
```diff
@@ -773,10 +800,9 @@ func (aws *AWS) ClusterManagementPricing() (string, float64, error) {
-	pricingURL := "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/"
+func getPricingListURL(serviceCode string, nodeList []*clustercache.Node) string {
+	// See https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/using-the-aws-price-list-bulk-api-fetching-price-list-files-manually.html
+	pricingURL := "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/" + serviceCode + "/current/"
```
refactor this to const

Tomerfi1210 replied:
done.

ameijer commented on pkg/cloud/aws/provider.go:
```diff
@@ -786,7 +812,7 @@ func (aws *AWS) getRegionPricing(nodeList []*clustercache.Node) (*http.Response,
 			// Switch to Chinese endpoint for regions with the Chinese prefix
 			if strings.HasPrefix(currentNodeRegion, "cn-") {
-				pricingURL = "https://pricing.cn-north-1.amazonaws.com.cn/offers/v1.0/cn/AmazonEC2/current/"
+				pricingURL = "https://pricing.cn-north-1.amazonaws.com.cn/offers/v1.0/cn/" + serviceCode + "/current/"
```
const

Tomerfi1210 replied:
done.

ameijer commented on pkg/cloud/aws/provider_test.go:
```diff
@@ -488,3 +488,308 @@ func Test_getStorageClassTypeFrom(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			aws := &AWS{
+				Clientset: &mockClusterCache{pods: tt.pods},
```
build failure here

Tomerfi1210 replied:
should be ok now, thanks

ameijer commented on pkg/cloud/aws/fargate_test.go:
```diff
@@ -0,0 +1,484 @@
+	defer server.Close()
+
+	// Set up test environment variable to use our test server
+	t.Setenv("AWS_ECS_PRICING_URL", server.URL)
```
interesting approach

ameijer replied:
I think I like it. should we consider some kind of a test that compares the test data to the actual AWS url to alert us when its format changes?

Tomerfi1210 replied:
Great idea, tell me if what I added is ok.

ameijer reviewed (commented):
thank you for the contribution @Tomerfi1210 ! had a few suggestions when you have a chance. very excited about this!

Tomerfi1210 commented:
> thank you for the contribution @Tomerfi1210 ! had a few suggestions when you have a chance. very excited about this!

Hi, sure. Im happy to contribute and help

Commit:
fix: Correct env function name and add missing mock method

- Fix GetAWSECSPricingURL() to GetAWSECSPricingURLOverride()
- Add GetAllResourceQuotas() to mockClusterCache to satisfy interface

Signed-off-by: tomerfi1210 <tomerfisher@macbookpro.local>

Commit:
refactor: Clean up AWS pricing code and standardize formatting

- Extract pricing URL constants for better maintainability
- Simplify getPricingListURL logic for China region handling
- Add nil check for FargatePricing in createFargateNode
- Standardize const declarations and code formatting
- Rename env var constant for consistency

Signed-off-by: tomerfi1210 <tomerfisher@macbookpro.local>

Commit:
test(aws): Add integration test to validate AWS pricing API format

Add TestFargatePricing_ValidateAWSPricingFormat to detect when AWS changes
their pricing API structure. Test is skipped by default and only runs with
INTEGRATION=true to avoid hitting AWS APIs in regular CI runs.

Signed-off-by: tomerfi1210 <tomerfisher@macbookpro.local>

Tomerfi1210 commented:
@ameijer Hi, thanks for the review, I fixed the code to what you comment.

IdoKendo commented:
Excited to see this get merged!

ameijer reviewed (approved):
LGTM, thanks @Tomerfi1210 !
